src/librustc/traits/error_reporting.rs

```diff
@@ -871,7 +871,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                                          obligation.param_env,
                                                          new_trait_ref.to_predicate());
 
-                    if selcx.evaluate_obligation(&new_obligation) {
+                    if selcx.evaluate_obligation(&new_obligation).must_apply() {
```

Comment:
Any reason this you didn't do `self.predicate_may_hold`, or was it just a quick fix? Does this also need intercrate mode by any chance?

Reply:
Well, I generally prefer not to give "helpful advice" unless we know it is correct, so I leaned towards `must_apply`, but `may_hold` might also be ok.

Reply:
Ah no I meant why didn't we use the canonical query here?